flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java

```diff
@@ -156,6 +157,9 @@
         this.numBytesOutCounter = metricGroup.getIOMetricGroup().getNumBytesOutCounter();
         this.numRecordsOutCounter = metricGroup.getIOMetricGroup().getNumRecordsOutCounter();
         this.numRecordsOutErrorsCounter = metricGroup.getNumRecordsOutErrorsCounter();
+        this.kafkaProducerConfig.setProperty(
+                ProducerConfig.CLIENT_ID_CONFIG,
+                createProducerClientId(kafkaProducerConfig, sinkInitContext.getSubtaskId()));
```

Comment:
It's possible to have multiple producers existing in the same writer. KafkaWriter has a producer pool for reusing, and creates a new producer once the pool is empty. I think there will be collision if those producers in the pool use the same client id.